Use the default template path

m-vo opened the pull request:
In 9a7ae410540b6a946e2c6a25d5fbbe1069049f14 (see #108) we set the default template path to `app/Resources/views`. This however prevents user from getting rid of the `app` folder if they are overwriting templates (e.g. our `service_unavailable.twig`). 

This PR resets the folder to the default (`/templates`) again and decorates the template iterator with one that chains the iterator with a filtering iterator that only allows `.twig` files. This circumvents the issue in #108 

todo: 
- [ ] check https://github.com/symfony/symfony/issues/34142
- [x] wait for response of PCT

Credits: This is a result of a discussion & debugging session together with @fritzmg. Thanks :smile:

m-vo committed:
twig: only cache template files with '.twig' extension

- decorate twig.template_iterator & add file filtering (see #108)
- configure symfony default template path (/templates)

m-vo commented:
(Not sure if we should target this against the 4.8. branch instead.)

leofeyer commented on manager-bundle/src/Twig/FileExtensionFilterIterator.php:
```diff
@@ -0,0 +1,46 @@
+     */
+    private function acceptsPath(string $path): bool
+    {
+        return 0 === strpos($path, '@') || self::FILE_EXTENSION === Path::getExtension($path);
```
`0 === strncmp($path, '@', 1)` could be more effective.

Toflar commented on manager-bundle/composer.json:
```diff
@@ -46,7 +46,8 @@
         "symfony/web-profiler-bundle": "4.2.* || 4.3.*",
         "symfony/yaml": "4.2.* || 4.3.*",
-        "toflar/psr6-symfony-http-cache-store": "^2.0"
+        "toflar/psr6-symfony-http-cache-store": "^2.0",
```
A new dependency just to check the file extension?🤔

leofeyer replied:
We already require `webmozart/path-util` in the core bundle, but I you certainly have a point here. Maybe we should just use `pathinfo()`, which the method does internally as well.

m-vo commented:
I updated the implementation and added tests. Imo this is ready besides a migration that moves the templates to the new spot. 

Where would we put this migration? In the manager-bundle tagged as a `contao.migration`? If so this depends on #709 I guess.

/ping @ausi

leofeyer reviewed (dismissed):
The `IteratorAggregateStub` should probably go into the `Fixtures` directory, otherwise 👍.

aschempp commented:
To fix BC, we should simply set the `twig.default_path` back to the `app/Resources/view` dir if the directory exists. Can be done in a compiler pass or maybe better in the manager plugin.

leofeyer commented:
I would prefer moving the files from `/app/Resources/views` into `/templates` during the version 4.9 update instead.

aschempp commented:
> I would prefer moving the files from `/app/Resources/views` into `/templates` during the version 4.9 update instead.

We don't move anything anywhere. We can't do that either because the files might be versionized locally. What if both folder exist or `app/...` is re-uploaded?

fritzmg commented:
The problem is the consistent documentation. We can't use both folders, thus we would only document: `app/Resources/...` for before Contao 4.9 and `templates/bundles/...` for Contao 4.9 and up.

aschempp commented:
Where's the difference to `app/Resources/config` and the root `/config` folder?

fritzmg commented:
> Where's the difference to `app/Resources/config` and the root `/config` folder?

The difference is that both work, where as with the template path only one of them works.

m-vo commented:
**Memo to myself**

Check if calls like this (but in the root namespace) could be affected: 
```twig
{{ include('@Twig/images/favicon.png.base64') }}
```
(example see [TwigBundle / layout.html.twig](https://github.com/symfony/symfony/blob/4045a134df3cfd2356b0b5a9fae7e91ec16d5f7d/src/Symfony/Bundle/TwigBundle/Resources/views/layout.html.twig#L10))

m-vo commented:
> We don't move anything anywhere. We can't do that either because the files might be versionized > locally. What if both folder exist or app/... is re-uploaded?

I would only move files if nothing exists at the destination - I think it's unlikely that anyone would use both paths? But yes, this is a question of deployment (and local versioning) vs. working directly on the server and I think you've got a point there. 🙃 

A compiler pass was my first intention as well. We would need to set the configuration before the Twig bundle reads it. Is this a matter of initialization order (via Manager Plugin) or how can this be controlled?

aschempp commented:
> A compiler pass was my first intention as well. We would need to set the configuration before the Twig bundle reads it. Is this a matter of initialization order (via Manager Plugin) or how can this be controlled?

A compiler pass is always before a service is run, so I don't think that would be a problem.

ausi commented:
As discussed in Mumble on October 24th, we should try to decorate the template loader of Symfony to support both paths. If that is not easily possible, we should use a compiler pass to set the path to `/app/Resources/views` if it exists and to `/templates` otherwise.

@m-vo suggested it might be possible to set `/app/Resources/views` as a templates path with an empty namespace as an alternative to decoration the templates loader.

We also want to invert the templates filter to match for `!== '.sql'` insteadof `=== '.twig'`.

m-vo commented:
> As discussed in Mumble on October 24th, we should try to decorate the template loader of Symfony to support both paths. 

The default `TemplateIterator` (found in the TwigBundle) always includes templates in `%kernel.root_dir%/Resources/views` (see [here](https://github.com/symfony/symfony/blob/591ad2203c90150fa4fdad82989d7a8a5060f3bb/src/Symfony/Bundle/TwigBundle/TemplateIterator.php#L54) for reference). 

With `app` beeing the default for `%kernel.root_dir%` this means the twig path we configured earlier (`%kernel.project_dir%/app/Resources/views`) was redundant ever since. I can remove the default path overwrite (like in this PR) and am still able to use the old path. In fact templates from the app folder are always given precedence over those of the default path.

This works for both application templates (e.g. `/app/Resources/views/demo.twig`) that users may use in their application as well as overwriting bundle configuration (e.g a custom `/app/Resources/ContaoCoreBundle/views/Error/service_unavailable.html.twig`).

**Conclusion:**
We can IMO change the path to the default again (like in this PR) without affecting anyone. Both paths should work. In case equally named templates exist at both places the legacy path will win however.

m-vo commented:
Btw.: Could anyone point me to the original Symfony issue where the concept of restricting twig templates to only files with a `.twig` extension was rejected? I don't think the current approach to compile all files in the cache warming stage is correct.

leofeyer commented:
> We can IMO change the path to the default again (like in this PR) without affecting anyone.

Haha, that's nice. 👍 

> Could anyone point me to the original Symfony issue where the concept of restricting twig templates to only files with a `.twig` extension was rejected?

I tried to find it during the Mumble call but I did not. You should probably ask in a new issue.

m-vo commented:
Let's keep an eye on the outcome of https://github.com/symfony/symfony/issues/34142 as well.

m-vo committed:
invert filtering logic and strip '.sql' files

fritzmg commented:
@ausi @m-vo PCT's database content files are _not_ in `.sql` files. They are in regular `.html5` files (or `.php`, I'll have to check).

The current implementation would not fix the original problem.

leofeyer commented:
@m-vo Is there anything I can do to help you complete this PR?

fritzmg commented:
Imho it's complete if https://github.com/contao/contao/pull/840/commits/360ebcc6602780d89b166a54b1ba8ac85eff8215 is reverted ;)

m-vo committed:
Revert "invert filtering logic and strip '.sql' files"

This reverts commit 360ebcc6602780d89b166a54b1ba8ac85eff8215.

m-vo committed:
move IteratorAggregateStub into tests\Fixtures

m-vo commented:
Sorry, I lost track of that one. :smiley:  Reverted 360ebcc6602780d89b166a54b1ba8ac85eff8215 as suggested by @fritzmg and moved the `IteratorAggregateStub` to `tests/Fixtures`. This should be ready now.

@leofeyer Do you want me to squash the commits?

m-vo commented:
I'm not sure why the PHP CS Fixer task reported an error here? 
It tries to fix `core-bundle/src/Resources/contao/library/Contao/DcaExtractor.php` which is not part of this PR.

leofeyer committed:
Fix the coding style

leofeyer commented on manager-bundle/src/Twig/FileExtensionFilterIterator.php:
```diff
@@ -0,0 +1,44 @@
+     */
+    private function acceptsPath(string $path): bool
+    {
+        return 0 === strncmp($path, '@', 1) || '.twig' === substr($path, -5);
```
Since this is only a one-liner, why did you encapsulate it in a separate method? You might as well move this into line 32. 🤷‍♂

m-vo replied:
Because then it's clear what this functional expression does: whether a path is accepted or not. This style is imo favorable over comments. :man_shrugging:

m-vo replied:
But it's up to you :smile:  If inlining is more consistent with the current code style, let's do that oc.

leofeyer reviewed (commented):
Why do we accept files that start with `@`?

m-vo commented:
> Why do we accept files that start with `@`?

Because imo we should not mess with namespaced templates (such as e.g. used by the Symfony profiler). If we really think it's worth filtering these as well, we can remove it, but I think there likely won't be any benefit but the risk to break things.

leofeyer commented:
But these templates are not stored as `templates/@template.html.twig`, are they?

m-vo commented:
No, their path rather looks like `@namespace/template.html.twig`.

leofeyer commented:
Exactly. And does our iterator not iterate over the file paths, which are actually stored in the file system? The condition will never be true then, will it?

m-vo commented:
No, the iterator will also contain everything from the bundle paths. Just put a `dump($path)` in there and run `cache:warm` and you'll see quite some files even in a 'naked' managed edition. 

see https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bundle/TwigBundle/TemplateIterator.php#L54-L87

m-vo commented on manager-bundle/src/Resources/config/services.yml:
```diff
@@ -48,3 +43,9 @@ services:
         arguments:
             - '@?contao_manager.jwt_manager'
+
+    contao_manager.twig.file_extension_filter_iterator:
```
What are the rules here? I thought we'd use FQCN for new definitions?

leofeyer replied:
No. FQCN are currently only used for  controllers. (This might change in the future though.)

m-vo replied:
I see.

m-vo replied:
We might want to define such CS rules somewhere... 

(@fritzmg Does a 'contributing' section belong to the docs? Or should this be in the repo?)

fritzmg replied:
There is a "Guides" section, however imho that should go into a CONTRIBUTING.md of the contao/contao repo, shouldn't it?

fritzmg replied:
> No. FQCN are currently only used for controllers. (This might change in the future though.)

But there also exist FQCN aliases for other services, not just for controllers.

leofeyer committed:
Make the callback filter iterator a static closure

leofeyer commented:
Thank you @m-vo.
